src/lib.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![warn(missing_docs)]
+
+//! List of well-known SS58 account types as an enum.
+use core::convert::TryFrom;
+
+include!(concat!(env!("OUT_DIR"), "/account_type_enum.rs"));
+
+/// Error encountered while parsing `Ss58AddressFormat` from &'_ str
+/// unit struct for now.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct ParseError;
+
+/// A custom address format. See also KnownSs58AddressFormat
+#[non_exhaustive]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct Ss58AddressFormat {
```

Comment:
Why did you split this out?

What is the reason?

Reply:
`Ss58AddressFormat::Custom(42)` should equal `Ss58AddressFormat::SubstrateAccount` but doesn't in the original code. An attacker might be able to take advantage of this somehow? Ideally we want to only be able to represent something in one way (for a particular type).

We could have custom implemented `PartialEq` but splitting them out has two additional advantages:
   * It uses half the memory - both Ss58AddressFormat and KnownSs58AddressFormat are now u16 size.
   * The cast from KnownSs58AddressFormat to Ss58AddressFormat is effectively just a mem transmute (before finding the enum's associated prefix required a match statement)